### ISSUE TITLE
[Needed 11/3/2020] Mezei spin flipper: modify flipper control

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#4871]
As an instrument scientist using the mezei spin flipper on LET (and POLREF), I would like IBEX to support the next iteration of the spin flipper hardware and software.

The spin flipper delivers a shaped current over time within a neutron pulse to some solenoid coils. Previously, the scientists had modelled the required shape of the current as an equation something like:

`I = A / (t + t0)`

Where `I` is current, `A` is an amplitude constant, `t` is the time within a neutron pulse, and `t0` is a time adjustment factor due to delays in the electronics. At present, IBEX sends these various constants to the flipper control software which is running on a remote computer.

After testing on LET, the scientists have realised that they cannot model the required current as a simple equation as above any more. Instead, they would now like to generate an `I` vs `t` calibration table. Instead of sending a small number of constants modelling a curve to the flipper control software, we would now like to send that calibration table.

The calibration table is likely to have ~10k rows. Since this is not enormous and will be sent infrequently, @davidkeymer and I suggested that sending the whole table over the TCP connection is likely to be fine. They are also happy to modify the flipper control software, running on the remote computer, to use this new protocol. I did stress to them that we are happy to provide advice on how to implement reliable communication code if needed.

The calibration table will be generated by an optimization routine that the scientists are also happy to write. This optimization routine will **minimize** the integrated intensity of a monitor between two wavelength/ToF bounds. It is a minimization because only unflipped neutrons are transmitted to the monitor. I have suggested that they should use `g.get_spectrum(...)` to get the appropriate monitor spectra - they may ask us for extra functionality in a separate ticket.

Goran will get in touch when he has implemented the relevant communications code. We will then need to modify the flipper IOC to use this new protocol.

Timescales for this change are an offline test in January and "online" use/testing in the Feb/Mar cycle.

**Comment by kjwoodsISIS** [comment]
10K rows for a calibration table seems rather a lot (excessive even?).  Are they sure they really need this number?  Just how much variation is there in the table?

**Comment by Tom-Willemsen** [comment]
@ChrisColeExpControl and I talked with Goran about this this afternoon. The setup is going to be used on LET on 11th March.

Goran has re-written the code on the "remote" end. It is now simpler than before from our perspective, and we only need to provide the following PVs for a single flipper (previously there were both polariser and analyser coils, now there are only one set):
- Current [setpoint and readback] - floating point
- Compensation [setpoint and readback] - floating point
- On/off [setpoint and readback] boolean

Longer term, we would also like to add:
- A filepath to a file mapping t (time) to I (current). [setpoint and readback of filepath].
  * The scientist plans to put this file on a network share - we have explained to them the caveats with this approach (network being unavailable etc). These files will only be being upgraded
- A "control mode" dictating whether the numeric parameters or the file are being used
(if these items are easy, we can implement them as part of this ticket, otherwise please split them into a separate ticket)

Additionally, the scientist would like help next week looking at an "optimisation" script which he has written to determine some of the parameters above. @ChrisColeExpControl and I agreed to look at this with him next week.

**Comment by Tom-Willemsen** [comment]
@ChrisColeExpControl and I met with the scientist to clarify some requirements this afternoon:

- The code is at `\\olympic\Babylon5\Public\goran\flipper\testserver_2.py`
- For the experiment on 11th March, we do not need to send the file over epics. Instead we will just send a file path (as a string). This might be either a local filepath on the flipper control PC or to a network share.
- I have edited comment above with further clarification of get/set values
- The command syntax for sets is (for example): `flipper_state=on`. We have asked for this to include a terminator, it does not currently have one.
  * As a reply to a set it echoes the entire command.
- The scientist has not currently implemented readbacks, but we have asked him to do so. Exact syntax to be confirmed.

**Comment by kjwoodsISIS** [comment]
How many points?

**Comment by DominicOram** [comment]
Can you update the documentation on the dev wiki to reflect these changes?

**Comment by DominicOram** [comment]
Specifically we're a bit unsure on what compensation means

**Comment by DominicOram** [comment]
Can you also document why you moved away from using the supplied python code as an emulator?

**Comment by DominicOram** [comment]
All looks good after review, mostly documentation and style. Would be happy for this to go on an instrument without rework.

**Comment by Tom-Willemsen** [comment]
- Docs added to wiki
- Added an explanation of compensation to the wiki - although the scientist would be best placed to explain this properly.
- We switched away from using the python code as an emulator as it was going to be too tricky to insert our mocks in relevant places to make it act in a sensible way - the new flipper code is quite tightly coupled to things like DaqMX. 
- @FreddieAkeroyd has found some old threads on tech talk about `PV$` vs `PV.VAL$` vs `PV.$` for long string support, and is going to inquire on tech-talk whether this is intended behaviour or a bug. In the meantime I don't think we can use long strings as most of our infrastructure wouldn't support `$` in pv names (which is required to be able to set a long string record). 
- I did try again but couldn't get waveform records to write directly to streamdevice. I've documented the workaround at https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Stream-Device-Tips-and-Tricks
- Rework comments on PRs done
